samples/RhythmGame/test/testLockFreeQueue.cpp

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2018 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include "utils/LockFreeQueue.h"
+
+/**
+ * Tests
+ * =====
+ *
+ * READ/POP:
+ *  - if queue is empty return false
+ *  - if queue has elements return true
+ *  - returns correct value once
+ *  - returns correct value more than once
+ *
+ * WRITE/PUSH:
+ *  - if queue is full return false
+ *  - if queue has space return true
+ *
+ * WRAP:
+ *  - read from full queue, then write one more element, then correct read
+ *
+ * PEEK:
+ *  - peek once
+ *  - peek twice
+ *
+ * SIZE:
+ *  - correct zero
+ *  - correct once
+ *  - correct many
+ *
+ * CANNOT TEST:
+ *  - read/writes correct after wraparound (takes too long to run)
+ *  - Constructing queue with non power of 2 capacity results in assert failure (needs death test)
+ *  - Constructing queue with signed type results in assert failure (needs death test)
+ *  - Multi-threaded read/writes (complicated to test)
```

Comment:
MT is essential to test here, otherwise there is a big risk of running into very confusing problems at run time only.

Reply:
Done